packages/debug/src/browser/debug-contribution-manager.ts

```diff
@@ -0,0 +1,247 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { DebugConfiguration } from '../common/debug-common';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import { Disposable } from '@theia/core/lib/common/disposable';
+import { DebugService, DebuggerDescription } from '../common/debug-service';
+import { IJSONSchema, IJSONSchemaSnippet } from '@theia/core/lib/common/json-schema';
+import { Emitter, Event } from '@theia/core/lib/common/event';
+import { DebugSessionFactory } from './debug-session-contribution';
+import { DebugSessionOptions } from './debug-session-options';
+import { DebugSession } from './debug-session';
+import { TerminalService } from '@theia/terminal/lib/browser/base/terminal-service';
+import { EditorManager } from '@theia/editor/lib/browser/editor-manager';
+import { BreakpointManager } from './breakpoint/breakpoint-manager';
+import { LabelProvider } from '@theia/core/lib/browser/label-provider';
+import { MessageClient } from '@theia/core/lib/common/message-service-protocol';
+import { OutputChannelManager, OutputChannel } from '@theia/output/lib/common/output-channel';
+import { DebugPreferences } from './debug-preferences';
+import { DebugSessionConnection } from './debug-session-connection';
+import { IWebSocket } from 'vscode-ws-jsonrpc/lib/socket/socket';
+import { DebugSessionContributionRegistry } from './debug-session-contribution-registry';
+
+/**
+ * Manages both extension and plugin debuggers contributions.
+ */
+@injectable()
+export class DebugContributionManager {
+    protected readonly pluginContributors = new Map<string, DebugPluginContributor>();
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+    @inject(DebugService)
+    protected readonly debugService: DebugService;
+    @inject(DebugSessionContributionRegistry)
+    protected readonly sessionContributionRegistry: DebugSessionContributionRegistry;
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+    @inject(BreakpointManager)
+    protected readonly breakpoints: BreakpointManager;
+    @inject(LabelProvider)
+    protected readonly labelProvider: LabelProvider;
+    @inject(MessageClient)
+    protected readonly messages: MessageClient;
+    @inject(OutputChannelManager)
+    protected readonly outputChannelManager: OutputChannelManager;
+    @inject(DebugPreferences)
+    protected readonly debugPreferences: DebugPreferences;
+
+    protected readonly onDidContributionAddEmitter = new Emitter<string>();
+    readonly onDidContributionAdd: Event<string> = this.onDidContributionAddEmitter.event;
```

Comment:
the pattern is `on[Will|Did]VerbNoun?` - https://code.visualstudio.com/docs/extensionAPI/patterns-and-principles#_events